syntax = "proto3";

package grpc;

enum UserRole {
  USER = 0;
  ADMIN = 1;
  MODERATOR = 2;
}

enum TokenType {
  BEARER = 0;
  REFRESH = 1;
}

message User {
  string username = 1;
  string role = 2;  // Changed from UserRole enum to string to return 'USER', 'ADMIN', 'MODERATOR'
  string id = 3;
  bool is_active = 4;
  string created_at = 5;
  string updated_at = 6;
}

message Token {
    string token = 1;
    TokenType token_type = 2;
    int64 expires_in = 3;
}

message ErrorResponse {
  string error_code = 1;
  string error_message = 2;
  string timestamp = 3;
}

message AuthRequest {
    string username = 1;
    string password = 2;
}

message AuthResponse {
  bool success = 1;
  string message = 2;
  string timestamp = 3;
  Token access_token = 4;
  Token refresh_token = 5;
  User user = 6;
}

message SuccessResponse{
    bool success = 1; 
    string message = 2; 
    string timestamp = 3;
}

message UserId {
    string id = 1;
}

message UsernameRequest {
    string username = 1;
}

message UserResponse {
    bool success = 1;
    string message = 2;
    string timestamp = 3;
    User user = 4;
}

message UsersResponse {
    bool success = 1;
    string message = 2;
    string timestamp = 3;
    repeated User users = 4;
}

message GetUserResponse {
    bool success = 1;
    User user = 2;
}

message ListUsersRequest {
    int32 limit = 1;
    int32 offset = 2;
    optional string username = 3;
    optional UserRole role = 4;
    optional bool is_active = 5;
}

message CreateUserRequest {
    string username = 1;
    string password = 2;
    optional UserRole role = 3;
    optional bool is_active = 4;
}

message UpdateUserRequest {
    string id = 1;
    optional string username = 2;
    optional string password = 3;
    optional UserRole role = 4;
    optional bool is_active = 5;
}

message ChangePasswordRequest {
    string current_password = 1; 
    string new_password = 2;
    Token token = 3;
}


service AuthService {
    rpc authentication(AuthRequest) returns (AuthResponse);
    rpc registrate(AuthRequest) returns (AuthResponse);
    rpc logout(Token) returns (SuccessResponse);
    rpc refresh_token(Token) returns (AuthResponse);
    rpc get_user(Token) returns (User); 
    rpc verify(Token) returns (SuccessResponse); 
    rpc change_password(ChangePasswordRequest) returns (SuccessResponse);
}

service UserService {
    rpc create_user(CreateUserRequest) returns (UserResponse);
    rpc get_user(UserId) returns (GetUserResponse);
    rpc get_user_by_username(UsernameRequest) returns (GetUserResponse);
    rpc update_user(UpdateUserRequest) returns (GetUserResponse);
    rpc delete_user(UserId) returns (SuccessResponse);
    rpc list_users(ListUsersRequest) returns (UsersResponse);
}